core/src/main/java/com/inrupt/client/core/DefaultClient.java

```diff
@@ -126,6 +128,25 @@ Request upgradeRequest(final Request request, final Credential token) {
         return builder.build();
     }
 
+    @SuppressWarnings("unchecked")
+    static <T, R extends Throwable> Response<T> recover(final Response<T> res, final Throwable err) throws R {
+        if (err != null) {
+            if (err instanceof ClientHttpException) {
+                final ClientHttpException ex = (ClientHttpException) err;
+                if (ex.getStatusCode() == UNAUTHORIZED) {
+                    return new NoBodyResponse<T>(ex.getUri(), ex.getStatusCode(), ex.getHeaders());
+                }
+            } else if (err.getCause() instanceof ClientHttpException) {
+                final ClientHttpException ex = (ClientHttpException) err.getCause();
+                if (ex.getStatusCode() == UNAUTHORIZED) {
+                    return new NoBodyResponse<T>(ex.getUri(), ex.getStatusCode(), ex.getHeaders());
+                }
+            }
+            throw (R) err;
+        }
+        return res;
+    }
```

Comment:
This new functionality is quite complex and branching.
Would it not make sense to add tests that highlight its behavior specifically?